services/core/shared/core/compat/sdk_v5/blockchainIndex.js

```diff
@@ -0,0 +1,512 @@
+		`;
+
+		const missingBlockRanges = await blocksDB.rawQuery(missingBlocksQueryStatement);
+
+		result = missingBlockRanges;
+	}
+
+	const logContent = result.map(o => `${o.from}-${o.to} (${o.to - o.from + 1} blocks)`);
+	logContent.forEach(o => logger.info(`Missing blocks in range: ${o}`));
+
+	return result;
+};
+
+const getLastFinalBlockHeight = async () => {
+	// Returns the highest finalized block available within the index
+	// If index empty, default lastIndexedHeight (alias for height) to blockIndexLowerRange
+	const blocksDB = await getBlocksIndex();
+
+	const [{ height: lastIndexedHeight } = {}] = await blocksDB.find({
+		sort: 'height:desc',
+		limit: 1,
+		isFinal: true,
+	}, ['height']);
+
+	return lastIndexedHeight;
+};
+
+const getNonFinalHeights = async () => {
+	const blocksDB = await getBlocksIndex();
+
+	const [{ height: lastIndexedHeight } = {}] = await blocksDB.find({
+		sort: 'height:asc',
+		limit: 5000,
+		isFinal: false,
+	}, ['height']);
+
+	return lastIndexedHeight || [];
+};
+
+const indexMissingBlocks = async () => {
+	const genesisHeight = await getGenesisHeight();
+	const currentHeight = await getCurrentHeight();
+
+	// Missing blocks are being checked during start
+	// By default they are checked from the blockchain's beginning
+	// It is possible to resume indexing from the last safe height
+	// Uncomment the line below to (slightly) increase performance during start
+	// const lastScheduledBlock = await getIndexVerifiedHeight() || genesisHeight;
+	const lastScheduledBlock = genesisHeight;
+	const minReqHeight = config.indexNumOfBlocks > 0
+		? currentHeight - config.indexNumOfBlocks : genesisHeight;
+
+	// Lowest and highest block heights expected to be indexed
+	const blockIndexHigherRange = currentHeight;
+	const blockIndexLowerRange = Math.max(minReqHeight, lastScheduledBlock);
+
+	// Retrieve the list of missing blocks
+	const missingBlockRanges = await findMissingBlocksInRange(
+		blockIndexLowerRange, blockIndexHigherRange);
+
+	// Start building the block index
+	try {
+		if (missingBlockRanges.length === 0) {
+			// Update 'indexVerifiedHeight' when no missing blocks are found
+			const indexVerifiedHeight = await getIndexVerifiedHeight();
+			await setIndexVerifiedHeight(Math.max(indexVerifiedHeight, blockIndexHigherRange));
+		} else {
+			for (let i = 0; i < missingBlockRanges.length; i++) {
+				const { from, to } = missingBlockRanges[i];
+
+				logger.info(`Attempting to cache missing blocks ${from}-${to} (${to - from + 1} blocks)`);
+				/* eslint-disable-next-line no-await-in-loop */
+				await buildIndex(from, to);
+				/* eslint-disable-next-line no-await-in-loop */
+				await setIndexVerifiedHeight(to + 1);
+			}
+		}
+	} catch (err) {
+		logger.warn(`Missed blocks indexing failed due to: ${err.message}`);
+	}
+};
+
+const updateNonFinalBlocks = async () => {
+	// TODO: This function always reports empty final heights
+	// Make sure it is correct
```

Comment:
This can be fixed by updating the following line inside of `mapRows` in `services/core/shared/indexdb/mysql.js`:

```
				if (val || val === 0 || val === false) row[o] = getValue(cast(val, schema[o].type));
```

Additionally, please consider defaulting `isFinal` to `false` in `services/core/shared/core/compat/sdk_v5/schema/blocks.js`.

Reply:
Updated!